pkg/apis/core/v1alpha1/uibutton_types.go

```diff
@@ -0,0 +1,131 @@
+/*
+Copyright 2020 The Tilt Dev Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"context"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+
+	"github.com/tilt-dev/tilt-apiserver/pkg/server/builder/resource"
+	"github.com/tilt-dev/tilt-apiserver/pkg/server/builder/resource/resourcestrategy"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// UIButton
+// +k8s:openapi-gen=true
+type UIButton struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   UIButtonSpec   `json:"spec,omitempty"`
+	Status UIButtonStatus `json:"status,omitempty"`
+}
+
+// UIButtonList
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type UIButtonList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	Items []UIButton `json:"items"`
+}
+
+// UIButtonSpec defines the desired state of UIButton
+type UIButtonSpec struct {
+	Location UIComponentLocation `json:"location"`
+
+	// Text to appear on the button itself
+	Text string `json:"text"`
+
+	// The command to run when clicked.
+	// Tilt will set some env vars when running the command:
+	// - $BUTTON_RESOURCE_NAME
+	//    If the button is on a resource, the name of the resource
+	// - $TEXTFIELD_${TEXTFIELD_NAME} (if in default namespace)
+	// - $TEXTFIELD_${TEXTFIELD_NAMESPACE}_${TEXTFIELD_NAME} (if not in default namespace)
+	//    If the button is in the same location as a TextField named 'foo'
+	//    Tilt will set $TEXTFIELD_FOO to that field's value. (name gets upper-cased)
+	Cmd CmdSpec `json:"cmd"`
```

Comment:
I'm interested in hearing more! Sorry if y'all have already discussed this to death and are sick of it. If you don't think it's worth discussing any more, I can just follow your lead. If you want to help bring me along and humor me a bit, I'd be grateful.

> should everything that triggers a Cmd have a spec of that Cmd

I was thinking along the lines of Deployments having PodTemplateSpecs, especially since we need to fill in the Cmd's env with resource name / textfield values.

> Here's the doc:
https://docs.google.com/document/d/1alWrU1YGFwHehwkdWTFXGPZphT1Kfsk7r5X0dMQgOtk/edit

Is that the right link? I don't see any mention of this tension, and the Cmd described in that doc doesn't have a RestartOnSpec or FileWatches.

> so i think the more idiomatic way to handle this would be to give RestartOn a reference to UIButton.

It seems strange for Cmd to know about every type that could trigger it. If nothing else, does this mean this idiom is unavailable to third-party extensions?

I'm also unsure restart semantics really make sense for a button. IIUC, the Cmd/FileWatch relationship is more for local_resource.serve_cmd, where there are long-running processes that we want to be restarted with no changes in spec. For the buttons, I was figuring we'd just disable the button while the process was running (or at least turn the button into an "abort" button"). This feels more like a Job, where we'll create a pod, finish the work, and be done with it.

Just to help my understanding, Is the plan to model local_resource's update_cmd as a dormant Cmd with a RestartOnSpec for its filewatches, and then its serve_cmd as a Cmd with a RestartOnSpec specifying the update_cmd's successful completion?

I'm also happy to have an offline conversation if that'd be more efficient / comfortable (though I'd like to then write some of this up).

Reply:
sure, here's more discussion with more links:
https://app.clubhouse.io/windmill/story/11532/document-how-to-build-entr

re: "It seems strange for Cmd to know about every type that could trigger it."

ah, but the solution you're proposing has the opposite problem: UIButton should know about every action it should trigger, and have a complete spec of each one.

The current thinking is that there can reasonably be a couple "base" trigger types, but potentially unbounded action types

Reply:
> The current thinking is that there can reasonably be a couple "base" trigger types, but potentially unbounded action types

I think this is the crux of it (though I will be honest - I waffle back and forth constantly because of the tradeoffs both ways).

Another consideration was giving action types more freedom for how to react to specific "triggers", which is arguably just an extension of the unbounded action types, i.e. there could be complex logic / more than one way to react to a trigger, such as Live Update.

FWIW I think we could also add support in the other direction to ease implementation burden, e.g. on file change, `FileWatch` reconciler could update a `tilt.dev/last-file-change` annotation on entities referenced via spec somehow.

Reply:
One complicating factor here is that we need some level of bidirectional knowledge - in the `Cmd::RestartOn` case, how does the button reconciler find the resource that reacted to it so that it can set up a watch and update its own version of state accordingly?

Reply:
i once heard this described as a "do you drive on the left side of the road or the right side of the road?" decisions. you could make a reasonable case for either, but a fair amount of infra is going to depend on the decision you make, and you probably don't want to try to do both.